src/inmanta/agent/forking_executor.py

```diff
@@ -440,61 +440,80 @@ def can_be_cleaned_up(self, retention_time: int) -> bool:
         return self.connection.last_used_at
 
-    async def force_stop(self, grace_time: float = inmanta.const.SHUTDOWN_GRACE_HARD) -> None:
-        """Stop by process close"""
+    async def force_stop(self, grace_time: float = inmanta.const.EXECUTOR_GRACE_HARD) -> None:
+        """
+        Stop by process close
+
+        This method will never raise an exeption, but log it instead.
+        """
         self.closing = True
         await asyncio.get_running_loop().run_in_executor(
             self.owner.thread_pool, functools.partial(self._force_stop, grace_time)
         )
 
     def _force_stop(self, grace_time: float) -> None:
+        """This method will never raise an exeption, but log it instead, as it is used as a finalizer"""
         if self.closed:
             return
         try:
-            self.process.terminate()
-            self.process.join(grace_time)
-            self.process.kill()
-            self.process.join()
-        except ValueError:
-            # Process already gone:
-            pass
-        self._set_closed()
-
-    def _set_closed(self) -> None:
-        # this code can be raced from the join call and the disconnect handler
-        # relying on the GIL to keep us safe
-        if not self.closed:
-            self.closing = True
-            self.closed = True
-            self.process.close()
-            self.owner._child_closed(self)
+            if self.process.exitcode is None:
+                # Running
+                self.process.join(grace_time)
 
-    async def join(self, timeout: float) -> None:
-        if self.closed:
-            return
-        try:
-            await asyncio.get_running_loop().run_in_executor(
-                self.owner.thread_pool, functools.partial(self.process.join, timeout)
-            )
             if self.process.exitcode is None:
                 LOGGER.warning(
                     "Executor for agent %s with id %s didn't stop after timeout of %d seconds. Killing it.",
                     self.executor_id.agent_name,
                     self.executor_id.identity(),
-                    timeout,
+                    grace_time,
                 )
+                # still running! Be a bit more firm
                 self.process.kill()
-                await asyncio.get_running_loop().run_in_executor(
-                    self.owner.thread_pool, functools.partial(self.process.join, timeout)
-                )
-            self._set_closed()
+                self.process.join()
+            # Mypy is not happy about this, but I see no other way
+            # Much of the internal logic of multi-processing depends on this flag
+            # Methods are prone to failing when called out-of-sequence and
+            #   we have two code paths leading here (join and connection lost)
+            # The exception they raise are undocumented and too generic to handle well
+            if not self.process._closed:
+                self.process.close()
```

Comment:
I don't understand this. How does the process get closed and how would it remain open?